### PR TITLE
Add DMA support for both I2C and SPI

### PR DESCRIPTION
**Pull request by TargetTriple**

### CODE REVIEW

File: ssd1306/ssd1306.c [hunk 1]
```diff
@@ -189,12 +176,26 @@ void ssd1306_UpdateScreen(void) {
...
     //  * 64px   ==  8 pages
     //  * 128px  ==  16 pages
-    for(uint8_t i = 0; i < SSD1306_HEIGHT/8; i++) {
-        ssd1306_WriteCommand(0xB0 + i); // Set the current RAM page address.
-        ssd1306_WriteCommand(0x00);
-        ssd1306_WriteCommand(0x10);
-        ssd1306_WriteData(&SSD1306_Buffer[SSD1306_WIDTH*i],SSD1306_WIDTH);
-    }
+#if defined(SSD1306_USE_SPI)
+	ssd1306_WriteCommand(0xB0); // Set the current RAM page address.
```
Comment (afiskon) [inline ssd1306/ssd1306.c:180]: Why did you remove `+i` ?

File: ssd1306/ssd1306.c [hunk 2]
```diff
@@ -189,12 +176,26 @@ void ssd1306_UpdateScreen(void) {
...
+#endif
+    HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_SET); // un-select OLED
+
+#elif defined(SSD1306_USE_I2C)
+#if defined(SSD1306_USE_DMA)
+    HAL_I2C_Mem_Write_DMA(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, SSD1306_Buffer, SSD1306_BUFFER_SIZE);
+#else
+    HAL_I2C_Mem_Write(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, SSD1306_Buffer, SSD1306_BUFFER_SIZE, HAL_MAX_DELAY);
+#endif
+#endif
```
Comment (afiskon) [inline ssd1306/ssd1306.c:198]: Wouldn't it be better to encapsulate these ifdefs into ssd1306_WriteData?

File: ssd1306/ssd1306.c [hunk 3]
```diff
@@ -189,12 +176,26 @@ void ssd1306_UpdateScreen(void) {
...
+#if defined(SSD1306_USE_SPI)
+	ssd1306_WriteCommand(0xB0); // Set the current RAM page address.
+    ssd1306_WriteCommand(0x00);
+    ssd1306_WriteCommand(0x10);
+	HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_RESET); // select OLED
+    HAL_GPIO_WritePin(SSD1306_DC_Port, SSD1306_DC_Pin, GPIO_PIN_SET); // data
+#if defined(SSD1306_USE_DMA)
+    HAL_SPI_Transmit_DMA(&SSD1306_SPI_PORT, SSD1306_Buffer, SSD1306_BUFFER_SIZE);
+#else
+    HAL_SPI_Transmit(&SSD1306_SPI_PORT, SSD1306_Buffer, SSD1306_BUFFER_SIZE, HAL_MAX_DELAY);
```
Comment (afiskon) [inline ssd1306/ssd1306.c:188]: What are the benefits of using HAL_SPI_Transmit_DMA vs HAL_SPI_Transmit here? Isn't the main advantage of DMA that you can transmit data asynchronously? This being said I don't have much experience with DMA...

**Review by afiskon — CHANGES_REQUESTED** [review]
See comments

File: ssd1306/ssd1306.c (hunk 1 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:180]: Because the addressing mode was initialized to be Horizontal Addressing Mode in ssd1306_Init().
That is, the address of the first column of the 2nd page is the address of the last column of the 1st page + 1.
So you needn't the for() loop to transfer the fram buffer one page by one page, just transfer all bytes one by one, as if there's only one page and its length is SSD1306_BUFFER_SIZE. I've tested OK.

File: ssd1306/ssd1306.c (hunk 2 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:198]: Yes, these macros can be placed in ssd1306_WriteData, which seems better, you're right. I'll do this.
Data is transferred. I tested my code on my STM32F103C8T6 under I2C with DMA, I2C without DMA, SPI with DMA and SPI without DMA with function ssd1306_TestFPS(). FPS is higher when DMA is enabled.

File: ssd1306/ssd1306.c (hunk 3 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:188]: Yes, DMA can transfer data asynchronously, and transfers much faster than without DMA. Using DMA also makes CPU not wasting time on moving data, so CPU have more time to do other work.

File: ssd1306/ssd1306.c (hunk 2 above)
Comment (r2axz) [reply ssd1306/ssd1306.c:204]: Unfortunately, DMA and non-DMA versions of SPI and I2C transmit functions are not interchangeable.

HAL_SPI_Transmit_DMA does not wait for the previous DMA transaction to complete. Therefore, if you call HAL_SPI_Transmit_DMA during an active SPI DMA transaction, it will return HAL_BUSY and won't send any data. Since this situation is not handled in the code, ssd1306_WriteData will occasionally lose data.

(see: [HAL_SPI_Transmit_DMA source code](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_spi.c#L1674))

HAL_I2C_Mem_Write_DMA has a different but equally dangerous pitfall. The function tries to wait until I2C is ready for the next DMA transfer. It does this by checking I2C_FLAG_BUSY in a loop. As soon as I2C becomes free, it starts a new DMA transfer. I2C_FLAG_BUSY is cleared in I2C_DMAXferCplt, which in turn is called by the DMA interrupt handler.

When you call HAL_I2C_Mem_Write_DMA from an interrupt handler, and the Interrupt Priority Group of the DMA interrupt is equal or less than the current interrupt priority group, you have a deadlock. The DMA interrupt will not be able to reset I2C_FLAG_BUSY because it cannot interrupt the interrupt handler being currently executed. This can be resolved by configuring proper interrupt priority grouping, but the PR does not have anything related to such configuration. This deadlock condition will freeze the firmware for HAL_MAX_DELAY ms, which practically means forever.

(see: [HAL_I2C_Mem_Write_DMA](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L3142))

Although this PR offers some degree of performance improvement, it also introduces the issues described above, which degrade stability.

Comment (TargetTriple) [reply ssd1306/ssd1306.c:204]: Really thanks for your detailed explianation! I'll fix. Learned a lot from you!

**Review by afiskon — CHANGES_REQUESTED** [review]
See comments

File: ssd1306/ssd1306.c [hunk 4]
```diff
@@ -36,6 +40,11 @@ void ssd1306_Reset(void) {
 void ssd1306_WriteCommand(uint8_t byte) {
     HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_RESET); // select OLED
     HAL_GPIO_WritePin(SSD1306_DC_Port, SSD1306_DC_Pin, GPIO_PIN_RESET); // command
+    while(SSD1306_SPI_PORT.State != HAL_SPI_STATE_READY); 	// Wait until previous transmit is done.
+                                                            // Before transmit a new sequence,
+                                                            // neither HAL_SPI_Transmit_DMA() nor HAL_SPI_Transmit() 
+                                                            // confirms the previous transmit is done in
+                                                            // STM32F1xx HAL driver V1.1.7, Oct. 20, 2020
     HAL_SPI_Transmit(&SSD1306_SPI_PORT, (uint8_t *) &byte, 1, HAL_MAX_DELAY);
```
Comment (afiskon) [inline ssd1306/ssd1306.c:48]: Shouldn't it be another ifdef here?

File: ssd1306/ssd1306.c [hunk 5]
```diff
@@ -189,12 +203,12 @@ void ssd1306_UpdateScreen(void) {
...
-        ssd1306_WriteCommand(0xB0 + i); // Set the current RAM page address.
-        ssd1306_WriteCommand(0x00);
-        ssd1306_WriteCommand(0x10);
-        ssd1306_WriteData(&SSD1306_Buffer[SSD1306_WIDTH*i],SSD1306_WIDTH);
-    }
+#if defined(SSD1306_USE_SPI)
+    ssd1306_WriteCommand(0xB0); // Set the current RAM page address.
+    ssd1306_WriteCommand(0x00);
+    ssd1306_WriteCommand(0x10);
+#endif
```
Comment (afiskon) [inline ssd1306/ssd1306.c:210]: Could you please submit this refactoring as a separate PR? Also, I don't understand why it's inside ifdef(DMA)...

File: ssd1306/ssd1306_conf_template.h [hunk 6]
```diff
@@ -17,6 +17,17 @@
...
 //#define STM32F7
 
+/*
+* Warning: If use DMA, you must enable I2C event interrupt or SPI global
+* interrupt, otherwise the hard fault interrupt will be triggered.
+*
+* If you use DMA and ssd1306_UpdateScreen() will be called in an interrupt
+* service routine (ISR), you must ensure the ISR can be interrupted by DMA
+* ISR for I2C or SPI, which unlocks I2C or SPI hardware, otherwise I2C or
+* SPI hardware will be busy forever.
```
Comment (afiskon) [inline ssd1306/ssd1306_conf_template.h:27]: I'm afraid this comment is not good enough. Could you please provide a step-by-step instructions on how to do it in STM32CubeMX and/or corresponding code examples? Better place it in a separate Markdown file and link it from the code.

File: ssd1306/ssd1306.c [hunk 7]
```diff
@@ -16,7 +16,11 @@ void ssd1306_WriteCommand(uint8_t byte) {
 
 // Send data
 void ssd1306_WriteData(uint8_t* buffer, size_t buff_size) {
+#if defined(SSD1306_USE_DMA)
+    HAL_I2C_Mem_Write_DMA(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, buffer, buff_size);
+#else
     HAL_I2C_Mem_Write(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, buffer, buff_size, HAL_MAX_DELAY);
+#endif
```
Comment (afiskon) [inline ssd1306/ssd1306.c:23]: Shouldn't corresponding _WriteCommand be changed as well?

**Review by afiskon — CHANGES_REQUESTED** [review]
Merging this PR may take a while since it is quite invasive and requires a thorough review and testing. Please address my comments.

**Comment by afiskon** [comment]
One more thing. Since this is an optimization it would be nice of you to provide benchmark tests you did to verify
- that the code got faster
- there was no performance degradation for non-DMA version
and corresponding results.

File: ssd1306/ssd1306.c (hunk 7 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:23]: `ssd1306_WriteCommand()` is used to transfer only one byte each time. If use DMA in `ssd1306_WriteCommand()`, CPU need to read the byte's address and transfer the address to DMA, then DMA use the address to transmit only one byte to I2C.
CPU still need to read every byte, DMA's effect isn't good here.
CPU transmit to DMA, then DMA transmit to I2C, maybe slower than CPU directly transmit to I2C.
But when transmit frame buffer, CPU only need to tell DMA start address and size of buffer, only 2 variables, then DMA can transmit these 128*64 bytes without CPU's help. DMA's effect is good here.

If want DMA to show its advantage, need to place all command bytes needed by an operate ( such as tens of the command bytes in `ssd1306_Init()` ) in an array, then use `HAL_I2C_Mem_Write_DMA()` only one time to transfer the whole array.

So, I prefer not using DMA in `ssd1306_WriteCommand()`.

File: ssd1306/ssd1306.c (hunk 4 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:48]: Same as [above](https://github.com/afiskon/stm32-ssd1306/pull/42/files/861f7e42293779e1272632e7110c7f1032a4085a#r654851563).

File: ssd1306/ssd1306.c (hunk 2 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:210]: In fact, neither `HAL_I2C_Mem_Write_DMA` nor `HAL_I2C_Mem_Write` waits for the previous transfer to be done.
[`HAL_I2C_Mem_Write_DMA` returned if `hi2cx->State != HAL_I2C_STATE_READY`.](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L3303)
[`HAL_I2C_Mem_Write` returned if `hi2cx->State != HAL_I2C_STATE_READY`.](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L2645)
Ah, I'd better manually use `while( hi2cx.State != HAL_I2C_STATE_READY )` to wait.

File: ssd1306/ssd1306_conf_template.h (hunk 6 above)
Comment (TargetTriple) [reply ssd1306/ssd1306_conf_template.h:27]: OK

**Comment by TargetTriple** [comment]
OK. It's my first time to PR in github. You're very patient and kind! Really thank you!

File: ssd1306/ssd1306.c (hunk 5 above)
Comment (TargetTriple) [reply ssd1306/ssd1306.c:210]: These three `ssd1306_WriteCommand()` is used to tell SSD1306 which address do you want to begin to write into. The three are necessary for SPI, but ont necessary for I2C.
Because I2C has function `HAL_I2C_Mem_Write_DMA()`, you can pass the GRAM start address to it, but I did not find such a function for SPI.
